### PR TITLE
fixes #18846 - Passes right params for AK content overrides

### PR DESCRIPTION
**Pull request by parthaa**

### CODE REVIEW

**Comment by parthaa** [comment]
Note this PR relies on -> https://github.com/Katello/katello/pull/6656

**Comment by akofink** [comment]
@parthaa Ping me once the Katello PR gets approved.

**Comment by parthaa** [comment]
Following series of hammer commands are made available with this change

- new _overrides_ section in  activation-key  product-content call to show more overrides. Its still limited by space but 99% of the times the user is only going to be interested in checking for enabled and more over the UI doesnt even let you set anything else other than enabled.  
```bash
$ hammer activation-key  product-content  --id=6
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|-----------------------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE              
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|-----------------------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | protected:1, enabled:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |                       
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |                       
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|-----------------------


```

- A way to remove a content override

```bash
$ hammer activation-key  content-override --remove --name=protected --id=6  --content-label=rhel-7-server-rpms
Updated content override

$ hammer activation-key  product-content  --id=6
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE 
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |          
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |          
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------
```


- A way to add a content override
```bash
$ hammer activation-key  content-override --value=1 --name=priority --id=6  --content-label=rhel-7-server-rpms
Updated content override

$ hammer activation-key  product-content  --id=6
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE             
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:1, priority:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |                      
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |                      
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
```

- Multiple ways to set _enabled_ on a content override. Note this going to be the most common use case. Infact 'name' is optional for enabled because the prior Katello API's assumed one was only setting the name. So basically the new api assumes that if you didnt pass a name you must be sending enabled.

```bash
$ hammer activation-key  content-override --value=1 --name=enabled --id=6  --content-label=rhel-7-server-rpms

$ hammer activation-key  content-override --value=yes --name=enabled --id=6  --content-label=rhel-7-server-rpms

#name optional
$ hammer activation-key  content-override --value=1 --id=6  --content-label=rhel-7-server-rpms

```

Same thing with content hosts. 

After registering host using the above activation key one can run the following commands to verify the overrides applied to a host.


```bash

$ hammer host subscription product-content --host-id=2 
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE             
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:1, priority:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |                      
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |                      
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------

```

Similarly one can add remove overrides on a per host basis

```bash
$ hammer host subscription content-override --host-id=2 --remove --name=priority --content-label=rhel-7-server-rpms
Updated content override
$ hammer host subscription product-content --host-id=2
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE 
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |          
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |          
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------


# Add the priority back

$ hammer  host subscription content-override --host-id=2 --value=1 --name=priority --content-label=rhel-7-server-rpms
Updated content override
$ hammer host subscription product-content --host-id=2
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE             
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:1, priority:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |                      
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |                      
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------


# Same logic with enabled as AK. "enabled" is inferred if name is blank

$ hammer  host subscription content-override --host-id=2 --value=0  --content-label=rhel-7-server-rpms
Updated content override
$ hammer host subscription product-content --host-id=2
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
ID   | NAME                                                     | TYPE | URL | GPG KEY | LABEL                          | ENABLED? | OVERRIDE             
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------
2456 | Red Hat Enterprise Linux 7 Server (RPMs)                 |      |     |         | rhel-7-server-rpms             | yes      | enabled:0, priority:1
168  | Red Hat Enterprise Linux 6 Server (RPMs)                 |      |     |         | rhel-6-server-rpms             | yes      |                      
5117 | Red Hat Virtualization 4 Agents for RHEL 6 Server (RPMs) |      |     |         | rhel-6-server-rhv-4-agent-rpms | no       |                      
-----|----------------------------------------------------------|------|-----|---------|--------------------------------|----------|----------------------


```

**Comment by parthaa** [comment]
@akofink should be good to test most of the items are in master. You need one small PR though (https://github.com/Katello/katello/pull/6706) to test ```hammer host subscription product-content --host-id=2``` . Rest should be readily available in master

**Comment by parthaa** [comment]
I am not sure why travis is failing for methods I didnt touch.

File: lib/hammer_cli_katello/activation_key.rb [hunk 1]
```diff
@@ -219,10 +231,44 @@ class ContentOverrideCommand < HammerCLIKatello::SingleResourceCommand
...
+          override = { 'content_label' => option_content_label }
+          override['value'] = option_value if option_value
+          override['remove'] = true if option_remove?
+          override['name'] = name
+          opts['content_overrides'] = [override]
+        end
+      end
+
+      build_options do |o|
+        o.expand.except(:content_overrides)
```
Comment (akofink) [inline lib/hammer_cli_katello/activation_key.rb]: Hmm... I don't think you need this line.

File: lib/hammer_cli_katello/host_subscription.rb [hunk 2]
```diff
@@ -92,6 +92,88 @@ def request_params
...
+          field :type, _("Type")
+          field :contentUrl, _("URL")
+          field :gpgUrl, _("GPG Key")
+          field :label, _("Label")
+        end
+        field :enabled, _("Enabled?"), Fields::Boolean
+        field :override_description, _("Override")
+      end
+
+      def extend_data(mod)
```
Comment (akofink) [inline lib/hammer_cli_katello/host_subscription.rb]: Maybe you should DRY up this code with a ContentOverrides module.

File: lib/hammer_cli_katello/host_subscription.rb [hunk 3]
```diff
@@ -92,6 +92,88 @@ def request_params
...
+          "#{override['name']}:#{value}"
+        end
+        mod["override_description"] = formatted_overrides.join(", ")
+        mod
+      end
+
+      build_options
+    end
+
+    class ContentOverrideCommand < HammerCLIKatello::SingleResourceCommand
```
Comment (akofink) [inline lib/hammer_cli_katello/host_subscription.rb]: Same thing here. DRY it up!

File: test/functional/activation_key/content_override_test.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
+require File.join(File.dirname(__FILE__), '../../test_helper')
+
+describe 'activation-key content-override' do
+  before do
+    @cmd = %w(activation-key content-override)
+  end
+  it "attach a content label" do
```
Comment (akofink) [inline test/functional/activation_key/content_override_test.rb]: s/attach/attaches/

Do this for all the tests.

File: test/functional/host/subscription/product_content_test.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+require File.join(File.dirname(__FILE__), '../../test_helper')
+require File.join(File.dirname(__FILE__), '../host_helpers')
```
Comment (akofink) [inline test/functional/host/subscription/product_content_test.rb]: Require the file under test as well.

File: test/functional/host/subscription/content_override_test.rb [hunk 6]
```diff
@@ -0,0 +1,94 @@
+require File.join(File.dirname(__FILE__), '../../test_helper')
+require File.join(File.dirname(__FILE__), '../host_helpers')
+
```
Comment (akofink) [inline test/functional/host/subscription/content_override_test.rb]: Require the file under test as well.

File: test/functional/activation_key/content_override_test.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
+require File.join(File.dirname(__FILE__), '../../test_helper')
+
```
Comment (akofink) [inline test/functional/activation_key/content_override_test.rb]: Require the file under test as well.

File: test/functional/host/subscription/product_content_test.rb (hunk 5 above)
Comment (parthaa) [inline test/functional/host/subscription/product_content_test.rb]: @akofink updated

**Comment by akofink** [comment]
LGTM! ACK!

**Pull request merged**